src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -232,9 +233,17 @@ int sd_dhcp_server_stop(sd_dhcp_server *server) {
         server->receive_message =
                 sd_event_source_unref(server->receive_message);
 
+        if (!server->bind_to_interface) {
+                server->receive_broadcast =
+                        sd_event_source_unref(server->receive_broadcast);
+        }
+
         server->fd_raw = safe_close(server->fd_raw);
         server->fd = safe_close(server->fd);
 
+        if (!server->bind_to_interface)
+                server->fd_broadcast = safe_close(server->fd_broadcast);
```

Comment:
as above, drop the if check, destroy the fd no matter what. safe_close() is a NOP if you pass < 0, to make this safe and easy